src/cpu/pred/simple_btb.hh

```diff
@@ -52,41 +56,89 @@ namespace gem5
 namespace branch_prediction
 {
 
-class SimpleBTB : public BranchTargetBuffer
+class BTBEntry : public CacheEntry
 {
   public:
-    SimpleBTB(const SimpleBTBParams &params);
 
-    void memInvalidate() override;
-    bool valid(ThreadID tid, Addr instPC) override;
-    const PCStateBase *lookup(ThreadID tid, Addr instPC,
-                           BranchType type = BranchType::NoBranch) override;
-    void update(ThreadID tid, Addr instPC, const PCStateBase &target_pc,
-                           BranchType type = BranchType::NoBranch,
-                           StaticInstPtr inst = nullptr) override;
-    const StaticInstPtr getInst(ThreadID tid, Addr instPC) override;
+    /** The entry's target. */
+    std::unique_ptr<PCStateBase> target;
 
+    /** The entry's thread id. */
+    ThreadID tid;
 
-  private:
-    struct BTBEntry
+    /** Pointer to the static branch instruction at this address */
+    StaticInstPtr inst;
+
+    /** Update the entry in the BTB */
+    void update(ThreadID _tid,
+                const PCStateBase &_target,
+                StaticInstPtr _inst)
     {
-        /** The entry's tag. */
-        Addr tag = 0;
+        tid = _tid;
+        set(target, _target);
+        inst = _inst;
+    }
 
-        /** The entry's target. */
-        std::unique_ptr<PCStateBase> target;
+    /** Match the tag of the BTB entry */
+    bool matchTag(const Addr _tag, ThreadID _tid)
+    {
+        return (matchTag(_tag) && (tid == _tid));
+    }
 
-        /** The entry's thread id. */
-        ThreadID tid;
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other) : CacheEntry(other)
+    {
+        tid   = other.tid;
+        inst  = other.inst;
+    }
+
+    /** Default constructor */
+    BTBEntry() : CacheEntry(), tid(0), inst(nullptr)  {}
 
-        /** Whether or not the entry is valid. */
-        bool valid = false;
+    using CacheEntry::matchTag;
+};
+
+class BTBCache : public AssociativeCache<BTBEntry>
+{
+  private:
+    BTBIndexingPolicy *indexingPolicy;
 
-        /** Pointer to the static branch instruction at this address */
-        StaticInstPtr inst = nullptr;
-    };
+  public:
+    BTBCache(const char *_my_name, const size_t num_entries,
+             const size_t associativity,
+             replacement_policy::Base *_replPolicy,
+             BaseIndexingPolicy *_indexingPolicy)
+        : AssociativeCache(_my_name, num_entries, associativity,
+                       _replPolicy, _indexingPolicy, BTBEntry())
+    {
+        indexingPolicy = static_cast<BTBIndexingPolicy*>(_indexingPolicy);
+    }
 
+    BTBEntry *findEntry(const Addr instPC, ThreadID tid) const;
+    BTBEntry *accessEntry(Addr instPC, ThreadID tid);
+    BTBEntry* findVictim(const Addr addr, ThreadID tid);
 
+  private:
+    using AssociativeCache<BTBEntry>::accessEntry;
```

Comment:
There are 6 functions in AssociativeCache´s public API that use Addr, yet only 3 of them are hidden away here. This BTB Cache uses the ThreadID as part of the address, if I understand correctly, so the users of this class should never be able to use those functions. The other 3 should be hidden away too.